File: shared/config/openai_client.py
"""Shared OpenAI-compatible client wrapper for Python labs.

Reads configuration from environment variables and provides a simple
chat.completions.create-style helper with structured logging and retries.
"""

import json
import logging
import os
from typing import Any

import httpx

logger = logging.getLogger("agent-labs")

DEFAULT_TIMEOUT = 60.0
DEFAULT_MAX_RETRIES = 3


def load_env(path: str | None = None) -> None:
    """Load .env file if python-dotenv is available."""
    try:
        from dotenv import load_dotenv

        load_dotenv(path or os.path.join(os.getcwd(), ".env"))
    except ImportError:
        pass


def get_config() -> dict[str, str]:
    load_env()
    return {
        "api_key": os.getenv("OPENAI_API_KEY", ""),
        "base_url": os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1").rstrip("/"),
        "model": os.getenv("OPENAI_MODEL", "gpt-4o-mini"),
        "log_level": os.getenv("LOG_LEVEL", "INFO"),
    }


def configure_logging(level: str | None = None) -> None:
    cfg = get_config()
    logging.basicConfig(
        level=getattr(logging, (level or cfg["log_level"]).upper(), logging.INFO),
        format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    )


class OpenAIClient:
    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        model: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ) -> None:
        cfg = get_config()
        self.api_key = api_key or cfg["api_key"]
        self.base_url = base_url or cfg["base_url"]
        self.model = model or cfg["model"]
        self.timeout = timeout
        self.max_retries = max_retries
        if not self.api_key and not self.base_url.startswith("http://localhost"):
            raise ValueError("OPENAI_API_KEY is required for non-local endpoints")

    def chat_completion(
        self,
        messages: list[dict[str, str]],
        model: str | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
        top_p: float | None = None,
        frequency_penalty: float | None = None,
        presence_penalty: float | None = None,
        stop: list[str] | str | None = None,
        seed: int | None = None,
        response_format: dict[str, str] | None = None,
        tools: list[dict[str, Any]] | None = None,
        tool_choice: str | dict[str, Any] | None = None,
        stream: bool = False,
        extra_body: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        url = f"{self.base_url}/chat/completions"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}",
        }
        payload: dict[str, Any] = {
            "model": model or self.model,
            "messages": messages,
            "stream": stream,
        }
        if temperature is not None:
            payload["temperature"] = temperature
        if max_tokens is not None:
            payload["max_tokens"] = max_tokens
        if top_p is not None:
            payload["top_p"] = top_p
        if frequency_penalty is not None:
            payload["frequency_penalty"] = frequency_penalty
        if presence_penalty is not None:
            payload["presence_penalty"] = presence_penalty
        if stop is not None:
            payload["stop"] = stop
        if seed is not None:
            payload["seed"] = seed
        if response_format is not None:
            payload["response_format"] = response_format
        if tools is not None:
            payload["tools"] = tools
        if tool_choice is not None:
            payload["tool_choice"] = tool_choice
        if extra_body:
            payload.update(extra_body)

        logger.debug("POST %s with payload: %s", url, json.dumps(payload, ensure_ascii=False))

        for attempt in range(1, self.max_retries + 1):
            try:
                with httpx.Client(timeout=self.timeout) as client:
                    response = client.post(url, headers=headers, json=payload)
                    response.raise_for_status()
                    return response.json()
            except (httpx.HTTPStatusError, httpx.NetworkError, httpx.TimeoutException) as exc:
                logger.warning("Request attempt %d failed: %s", attempt, exc)
                if attempt == self.max_retries:
                    raise
        raise RuntimeError("Unreachable")

    def extract_message(self, response: dict[str, Any]) -> dict[str, Any]:
        return response["choices"][0]["message"]


if __name__ == "__main__":
    configure_logging()
    client = OpenAIClient()
    resp = client.chat_completion(
        messages=[{"role": "user", "content": "Say hello in one word."}],
        max_tokens=10,
    )
    print(resp["choices"][0]["message"]["content"])

File: shared/config/openai_client.ts
// Shared OpenAI-compatible client wrapper for Node/TypeScript labs.
// Reads configuration from environment variables.

import * as dotenv from "dotenv";
import fetch from "node-fetch";

export interface OpenAIConfig {
  apiKey: string;
  baseUrl: string;
  model: string;
  logLevel: string;
}

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  name?: string;
  tool_call_id?: string;
}

export interface ChatCompletionOptions {
  model?: string;
  messages: ChatMessage[];
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  stop?: string[] | string;
  seed?: number;
  response_format?: Record<string, string>;
  tools?: unknown[];
  tool_choice?: string | Record<string, unknown>;
  stream?: boolean;
  n?: number;
  extra_body?: Record<string, unknown>;
}

export interface ChatCompletionResponse {
  choices: Array<{
    message: ChatMessage & { tool_calls?: unknown[] };
    finish_reason: string;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

export function loadConfig(path?: string): OpenAIConfig {
  dotenv.config({ path: path || ".env" });
  return {
    apiKey: process.env.OPENAI_API_KEY || "",
    baseUrl: (process.env.OPENAI_BASE_URL || "https://api.openai.com/v1").replace(/\/$/, ""),
    model: process.env.OPENAI_MODEL || "gpt-4o-mini",
    logLevel: process.env.LOG_LEVEL || "INFO",
  };
}

export class OpenAIClient {
  public readonly config: OpenAIConfig;
  private timeoutMs: number;
  private maxRetries: number;

  constructor(options?: Partial<OpenAIConfig & { timeoutMs?: number; maxRetries?: number }>) {
    const env = loadConfig();
    this.config = {
      apiKey: options?.apiKey ?? env.apiKey,
      baseUrl: options?.baseUrl ?? env.baseUrl,
      model: options?.model ?? env.model,
      logLevel: options?.logLevel ?? env.logLevel,
    };
    this.timeoutMs = options?.timeoutMs ?? 60000;
    this.maxRetries = options?.maxRetries ?? 3;
    if (!this.config.apiKey && !this.config.baseUrl.startsWith("http://localhost")) {
      throw new Error("OPENAI_API_KEY is required for non-local endpoints");
    }
  }

  async chatCompletion(options: ChatCompletionOptions): Promise<ChatCompletionResponse> {
    const url = `${this.config.baseUrl}/chat/completions`;
    const payload: Record<string, unknown> = {
      model: options.model ?? this.config.model,
      messages: options.messages,
      stream: options.stream ?? false,
    };
    if (options.temperature !== undefined) payload.temperature = options.temperature;
    if (options.max_tokens !== undefined) payload.max_tokens = options.max_tokens;
    if (options.top_p !== undefined) payload.top_p = options.top_p;
    if (options.frequency_penalty !== undefined) payload.frequency_penalty = options.frequency_penalty;
    if (options.presence_penalty !== undefined) payload.presence_penalty = options.presence_penalty;
    if (options.stop !== undefined) payload.stop = options.stop;
    if (options.seed !== undefined) payload.seed = options.seed;
    if (options.response_format !== undefined) payload.response_format = options.response_format;
    if (options.tools !== undefined) payload.tools = options.tools;
    if (options.tool_choice !== undefined) payload.tool_choice = options.tool_choice;
    if (options.n !== undefined) payload.n = options.n;
    if (options.extra_body) Object.assign(payload, options.extra_body);

    if (this.config.logLevel === "DEBUG") {
      console.debug("POST", url, JSON.stringify(payload, null, 2));
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.config.apiKey}`,
    };

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this.timeoutMs);
        const response = await fetch(url, {
          method: "POST",
          headers,
          body: JSON.stringify(payload),
          signal: controller.signal,
        });
        clearTimeout(timer);
        if (!response.ok) {
          const text = await response.text();
          throw new Error(`HTTP ${response.status}: ${text}`);
        }
        return (await response.json()) as ChatCompletionResponse;
      } catch (err) {
        console.warn(`Request attempt ${attempt} failed:`, err);
        if (attempt === this.maxRetries) throw err;
      }
    }
    throw new Error("Unreachable");
  }

  extractMessage(response: ChatCompletionResponse): ChatCompletionResponse["choices"][0]["message"] {
    return response.choices[0].message;
  }
}

async function main() {
  const client = new OpenAIClient();
  const resp = await client.chatCompletion({
    messages: [{ role: "user", content: "Say hello in one word." }],
    max_tokens: 10,
  });
  console.log(resp.choices[0].message.content);
}

if (require.main === module) {
  main().catch(console.error);
}
